pinot-plugins/pinot-stream-ingestion/pinot-kafka-2.0/src/main/java/org/apache/pinot/plugin/stream/kafka20/KafkaSSLUtils.java

```diff
@@ -0,0 +1,340 @@
+      );
+      KeyStore.PasswordProtection keyPasswordProtection = new KeyStore.PasswordProtection(keyPassword.toCharArray());
+      keyStore.setEntry(DEFAULT_CLIENT_ALIAS, privateKeyEntry, keyPasswordProtection);
+
+      // Save the keystore to the specified location
+      try (FileOutputStream fos = new FileOutputStream(keyStorePath.toString())) {
+        keyStore.store(fos, keyStorePassword.toCharArray());
+      }
+      LOGGER.info("Initialized the SSL key store.");
+    } catch (Exception ex) {
+      throw new RuntimeException("Error initializing the SSL key store", ex);
+    }
+  }
+
+  private static Path getTrustStorePath(Properties consumerProps) {
+    String trustStoreLocation = consumerProps.getProperty(SSL_TRUSTSTORE_LOCATION);
+    return Paths.get(trustStoreLocation);
+  }
+
+  private static Path getKeyStorePath(Properties consumerProps) {
+    String keyStoreLocation = consumerProps.getProperty(SSL_KEYSTORE_LOCATION);
+    return Paths.get(keyStoreLocation);
+  }
+
+  // Renew the trust store if needed.
+  private static void renewTrustStore(Properties consumerProps) {
+    boolean renewTrustStore;
+    Path trustStorePath = getTrustStorePath(consumerProps);
+    String trustStorePassword = consumerProps.getProperty(SSL_TRUSTSTORE_PASSWORD);
+    String serverCertificate = consumerProps.getProperty(STREAM_KAFKA_SSL_SERVER_CERTIFICATE);
+    String certificateType = consumerProps.getProperty(STREAM_KAFKA_SSL_CERTIFICATE_TYPE, DEFAULT_CERTIFICATE_TYPE);
+
+    try {
+      // Load the trust store
+      KeyStore trustStore = KeyStore.getInstance(KeyStore.getDefaultType());
+      try (FileInputStream fis = new FileInputStream(trustStorePath.toString())) {
+        trustStore.load(fis, trustStorePassword.toCharArray());
+      }
+
+      // Decode the provided certificate
+      byte[] decodedCertBytes = Base64.getDecoder().decode(serverCertificate);
+      CertificateFactory certFactory = CertificateFactory.getInstance(certificateType);
+      Certificate providedCertificate = certFactory.generateCertificate(new ByteArrayInputStream(decodedCertBytes));
+
+      // Get the certificate from the trust store
+      Certificate trustStoreCertificate = trustStore.getCertificate(DEFAULT_SERVER_ALIAS);
+
+      // Compare the certificates
+      renewTrustStore = !providedCertificate.equals(trustStoreCertificate);
+    } catch (FileNotFoundException fex) {
+      // create the trust store if trust store does not exist – happens the very first time
+      renewTrustStore = true;
+    } catch (Exception ex) {
+      // renew trust store if comparison check fails
+      renewTrustStore = true;
+      LOGGER.warn("Trust store certificate comparison check failed.", ex);
+    }
+
+    if (renewTrustStore && Files.exists(trustStorePath)) {
+      LOGGER.info("Renewing the SSL trust store");
+      deleteFile(trustStorePath);
+    }
+  }
+
+  // Renew the key store if needed.
+  private static void renewKeyStore(Properties consumerProps) {
```

Comment:
I think that the current code will delete the certificate file if `STREAM_KAFKA_SSL_CLIENT_CERTIFICATE` is null or empty while the other config.

Are we keeping backward compatibility for tables who already have TLS configured manual way? (having trustore/key store copied and the table config includes the location & password).

Reply:
Good point. I have enhanced the check to do the SSL initialization only when the certificates are explicitly passed in.